### PR TITLE
indexer-alt: event indices

### PR DESCRIPTION
**Pull request by amnn**
## Description 

This PR includes three changes (separated across three commits). The first one fixes some edge cases in the committer related to handling checkpoint streams that are either empty themselves, or produce no data to write, the second introduces the event index tables roughly as they are in the current schema, and the last commit suggests a simplification.

The main changes to the schema are as follows:

- When filters correspond to different levels of nesting in a cascading type/module filter, they are all held in a single table, with all the cascading filter's fields, and separate indices are added to this table to deal with each level of nesting.
  - We need to confirm that the reader side can handle this change as well.
- They do not include an event sequence number: They will be used to fetch the transaction that emitted the event in question, and to fetch the relevant events, we will need to scan through that transactions events (which will be bounded by the protocol limit on number of events emitted).
- The struct instantiation table stores the instantiation information (name and type params) in two separate fields, with the type params stored as a BCS encoded array of `TypeTag`. This is to keep footprint down as the textual representation of a type can be up to twice as large as the BCS representation.
- Names have been standardised.

## Test plan 

Run the indexer twice on the first 1000 checkpoints and using only the event pipelines. This produces no information on the first round, but updates the watermark, and on the second iteration, it will iterate through no checkpoints before shutting down:

```
sui$ cargo run -p sui-indexer-alt --                                             \
  --database-url "postgres://postgres:postgrespw@localhost:5432/sui_indexer_alt" \
  --remote-store-url https://checkpoints.mainnet.sui.io                          \
  --last-checkpoint 1000                                                         \
  --pipeline ev_emit_mod --pipeline ev_struct_inst
```

Then run the same pipelines over checkpoints 9,000,000 to 9,002,000, which do include events:

```
sui$ cargo run -p sui-indexer-alt --                                             \
  --database-url "postgres://postgres:postgrespw@localhost:5432/sui_indexer_alt" \
  --remote-store-url https://checkpoints.mainnet.sui.io                          \
  --first-checkpoint 9000000 --last-checkpoint 9002000                           \
  --pipeline ev_emit_mod --pipeline ev_struct_inst
```

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: crates/sui-indexer-alt/src/handlers/mod.rs
Comment (amnn) [reply crates/sui-indexer-alt/src/handlers/mod.rs:358]: This didn't end up entirely fixing the issue because the receiver can be closed after the `select` starts waiting, which will still cause all of the arms to be unresponsive. In general, it's probably a good rule of thumb to not include conditions on guard arms that can be triggered externally (in this case `rx.is_closed()` is one of those), because they can be have unpredictably.

The fix was to remove all guards from the `poll.tick()` case.

**Comment by amnn** [comment]
> changed schema as well (I will double check and come back, both jsonrpc and graphql.)

I wasn't expecting that we would port the read side over just yet, but instead that we would focus on the write side and getting something working end-to-end that we can stress test for write performance. Instead of making the read changes on the actual readers, maybe you could write updated queries and `EXPLAIN ANALYZE` those to confirm that they operate the same way? That would be a lower friction way to check the schema change does what we want.

Also do you think we could land the changes to the alternate indexer and fix forward if we need to change the schema? Given we don't have any DBs that use this schema yet, there does not seem much harm in doing that.

File: crates/sui-indexer-alt/src/handlers/mod.rs
Comment (gegaowp) [inline crates/sui-indexer-alt/src/handlers/mod.rs:641]: the back pressure logic here seems off, if `pending_rows >= H::MAX_PENDING_SIZE` and there are values in `rx`, the next value from `rx` will be consumed and then ignored?

File: crates/sui-indexer-alt/migrations/2024-10-19-113135_ev_indices/up.sql
Comment (lxfind) [inline crates/sui-indexer-alt/migrations/2024-10-19-113135_ev_indices/up.sql:6]: Any reason that sender is more special than other fields? It's the only NOT NULL field.

File: crates/sui-indexer-alt/src/handlers/mod.rs
Comment (gegaowp) [reply crates/sui-indexer-alt/src/handlers/mod.rs:641]: I tried to write a quick test to verify, but seems that the committer back pressure test will be good to have TempDB supported first, I am happy to help with the TempDB support and committer tests if you have not started on that yet

**Review by gegaowp — APPROVED** [review]
the changes of this pr look good, left a comment about committer back pressure

**Comment by emmazzz** [comment]
> Also do you think we could land the changes to the alternate indexer and fix forward if we need to change the schema?

For sure this concern is not blocking. And if we want to split into multiple tables later, we can just backfill those tables from the main one. 

> maybe you could write updated queries and EXPLAIN ANALYZE those to confirm that they operate the same way?

Yes this would be a good sanity check. To be clear, I don't mean that we need to have graphql benchmark working before we can land these changes. The reader side changes, if we want to implement them, can be done on the side (graphql-alt?) without blocking indexer-alt. Just thinking that we should keep this in mind while making the changes especially since the existence of these separate indices is entirely for graphql query performance.

**Review by emmazzz — APPROVED** [review]

File: crates/sui-indexer-alt/migrations/2024-10-19-113135_ev_indices/up.sql
Comment (amnn) [reply crates/sui-indexer-alt/migrations/2024-10-19-113135_ev_indices/up.sql:6]: The other fields are implicitly not null because they are part of the primary key.

File: crates/sui-indexer-alt/src/handlers/mod.rs
Comment (amnn) [reply crates/sui-indexer-alt/src/handlers/mod.rs:641]: The core parts of the indexer are changing quite a bit at the moment, so I would hold off on that for just now, otherwise we'll end up conflicting with each other a lot, but we can talk about ways to get involved in the sync today!

**Pull request merged**
